controllers/machine_controller.go

```diff
@@ -242,6 +256,73 @@ func (r *MachineReconciler) isDeleteNodeAllowed(ctx context.Context, machine *cl
 	}
 }
 
+func (r *MachineReconciler) drainNode(ctx context.Context, cluster *clusterv1.Cluster, nodeName string, machineName string) error {
+	var kubeClient kubernetes.Interface
+	if cluster == nil {
+		var err error
+		kubeClient, err = kubernetes.NewForConfig(r.config)
+		if err != nil {
+			return errors.Errorf("unable to build kube client: %v", err)
+		}
+	} else {
+		// Otherwise, proceed to get the remote cluster client and get the Node.
+		remoteClient, err := remote.NewClusterClient(r.Client, cluster)
+		if err != nil {
+			klog.Errorf("Error creating a remote client for cluster %q while deleting Machine %q, won't retry: %v",
+				cluster.Name, nodeName, err)
+			return nil
+		}
+		var err2 error
+		kubeClient, err2 = kubernetes.NewForConfig(remoteClient.RESTConfig())
+		if err2 != nil {
+			klog.Errorf("Error creating a remote client for cluster %q while deleting Machine %q, won't retry: %v",
+				cluster.Name, nodeName, err)
+			return nil
+		}
+	}
+
+	node, err := kubeClient.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If an admin deletes the node directly, we'll end up here.
+			klog.Infof("Machine %v: Could not find node %v from noderef, it may have already been deleted: %v", machineName, nodeName, cluster.Name)
+			return nil
+		}
+		return errors.Errorf("unable to get node %q: %v", nodeName, err)
+	}
+
+	drainer := &kubedrain.Helper{
+		Client:              kubeClient,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteLocalData:     true,
+		GracePeriodSeconds:  -1,
+		// If a pod is not evicted in 20 second, retry the eviction next time the
+		// machine gets reconciled again (to allow other machines to be reconciled)
+		Timeout:               20 * time.Second,
+		OnPodDeletedOrEvicted: onPodDeletedOrEvicted,
+		Out:                   writer{klog.Info},
+		ErrOut:                writer{klog.Error},
+		DryRun:                false,
+	}
+
+	if err := kubedrain.RunCordonOrUncordon(drainer, node, true); err != nil {
+		// Machine still tries to terminate after drain failure
```

Comment:
```suggestion
		// Machine still tries to terminate after cordon failure
```

Reply:
Is this comment accurate? Maybe something more along the lines of:
```
// Machine will be re-reconciled after a cordon failure
```
would be more accurate, since we are returning any errors returned from `drainNode` back to controller-runtime.